### PR TITLE
*: define error code for errors

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve?

Problem Summary:
Define error code for errors.

### What is changed and how it works?
Define error code for errors to provide a unified way to query the reason and workarounds for errors.

Proposal: [xxx](url) 

What's Changed:
* Introduce `ErrorCodeExt`
* Define error codes
* Impl `ErrorCodeExt` for errors

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
* Define error code for errors

### CODE REVIEW

File: components/tikv_util/src/error_code/codec.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+use super::ErrorCodeExt;
+use crate::codec::Error;
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use codec::Error as CodecError;
+
+define_error_codes!(
+    "KV:CODEC:",
```
Comment (sticnarf) [inline components/tikv_util/src/error_code/codec.rs:8]: Why is `CODEC` all in upper case?

File: components/tikv_util/src/error_code/codec.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+use super::ErrorCodeExt;
+use crate::codec::Error;
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (sticnarf) [inline components/tikv_util/src/error_code/codec.rs:3]: Put it at the very top of the file.

File: components/tikv_util/src/error_code/engine.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+define_error_codes!(
+    "KV:Engine:",
+
+    ENGINE => ("Engine", "", ""),
+    NOT_IN_RANGE => ("NotInRange", "", ""),
+    PROTOBUF => ("Protobuf", "", ""),
+    IO => ("IO", "", ""),
+    CFNAME => ("CFName", "", ""),
```
Comment (sticnarf) [inline components/tikv_util/src/error_code/engine.rs:10]: Prefer `CF_NAME`

File: components/tikv_util/src/error_code/engine.rs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    CODEC => ("Codec", "", ""),
+
+    UNDETERMINED => ("Undetermined", "", "")
```
Comment (sticnarf) [inline components/tikv_util/src/error_code/engine.rs:13]: I'm worried if this "undetermined error" will be messed up with transaction's undetermined error (when there is a network error during committing the primary key).

File: src/storage/mvcc/mod.rs [hunk 5]
```diff
@@ -250,6 +251,35 @@ impl From<txn_types::Error> for ErrorInner {
...
+    fn error_code(&self) -> ErrorCode {
+        match self.0.as_ref() {
+            ErrorInner::Engine(e) => e.error_code(),
+            ErrorInner::Io(_) => error_code::storage::IO,
+            ErrorInner::Codec(e) => e.error_code(),
+            ErrorInner::KeyIsLocked(_) => error_code::storage::KEY_IS_LOCKED,
+            ErrorInner::BadFormat(e) => e.error_code(),
+            ErrorInner::Committed { .. } => error_code::storage::COMMITTED,
+            ErrorInner::PessimisticLockRolledBack { .. } => {
+                error_code::storage::PESSIMISTIC_LOCK_ROLLEDBACK
```
Comment (sticnarf) [inline src/storage/mvcc/mod.rs:264]:
```suggestion
                error_code::storage::PESSIMISTIC_LOCK_ROLLED_BACK
```

File: components/tikv_util/src/error_code/sst_importer.rs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+define_error_codes!(
+    "KV:SSTImporter:",
+
+
```
Comment (sticnarf) [inline components/tikv_util/src/error_code/sst_importer.rs:6]: two blank lines?

**Comment by sticnarf** [comment]
Defining all error codes in one place may help to manager all errors. But it may make more trouble when people are adding new error variants. I don't know if there is a better solution.

File: components/tidb_query_common/src/error.rs [hunk 7]
```diff
@@ -106,3 +120,12 @@ impl<T: Into<EvaluateError>> From<T> for Error {
 }
 
 pub type Result<T> = std::result::Result<T, Error>;
+
+impl ErrorCodeExt for Error {
+    fn error_code(&self) -> ErrorCode {
+        match self.0.as_ref() {
+            ErrorInner::Storage(_) => error_code::coprocessor::STORAGE_ERROR,
```
[the remark this answers is not in the document]
Comment (5kbpers) [reply components/tidb_query_common/src/error.rs:127]: We can not match error code from `failure::Error` for now.

**Comment by 5kbpers** [comment]
> Defining all error codes in one place may help to manager all errors. But it may make more trouble when people are adding new error variants. I don't know if there is a better solution.

Me either. But IMO it just adds a little job to do.

**Pull request merged**